book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+complexities are connected together, the default value specified in the table
+below must be driven for any omitted signals.
+
+| Name    | Condition                                   | Default   |
+|---------|---------------------------------------------|-----------|
+| `valid` | \\(\top\\)                                  | `'1'`     |
+| `ready` | \\(\top\\)                                  | `'1'`     |
+| `data`  | \\(E > 0\\)                                 | all `'0'` |
+| `last`  | \\(D \ge 1\\)                               | all `'1'` |
+| `stai`  | \\(C \ge 6 \wedge N > 1\\)                  | 0         |
+| `endi`  | \\((C \ge 5 \vee D \ge 1) \wedge N > 1\\)   | \\(N-1\\) |
+| `strb`  | \\(C \ge 7 \vee D \ge 1\\)                  | all `'1'` |
+| `user`  | \\(U > 0\\)                                 | all `'0'` |
+
+### Interface conventions
+
+When there are name conflicts (for instance due to multiple physical streams
+existing on the hierarchical level of an interface) the signal names must be
+prefixed by a unique stream name, delimited using an underscore.
+
+The canonical representation for the `data` and `user` signals is the LSB-first
+concatenation of the contained fields. For \\(N > 1\\), the lanes are
+concatenated LSB first, such that the lane index is major and the field index
+is minor.
+
+Where applicable, the signals must be listed in the following order: `dn`,
+`up`, `valid`, `ready`, `data`, `last`, `stai`, `endi`, `strb`, `user`.
+
+> This is mostly just a consistency thing, primarily because it helps to get
+> used to a single ordering when interpreting simulation waveforms. It may also
+> simplify tooling.
+
+The signal names must be lowercase.
+
+> This is for interoperability between languages that differ in case
+> sensitivity.
+
+#### Alternative representation
+
+To improve code readability in hardware definition languages supporting array
+and aggregate constructs (record, struct, ...), the following changes are
+permissible. However, it is recommended to fall back to the conventions above
+for interoperability with other streamlets on the "outer" interfaces of an IP
+block.
+
+ - `valid`, `data`, `last`, `stai`, `endi`, `strb`, and `user` may be bundled
+   in an aggregate type named `<stream-name>_dn_type` with signal name
+   `<stream-name>_dn`.
+
+ - `ready` may be "bundled" in an aggregate type named `<stream-name>_up_type`
+   with signal name `<stream-name>_up` for symmetry.
+
+ - If the language allows for signal direction reversal within a bundle, all
+   stream signals may also be bundled into a single type named
+   `<stream-name>_type`, with `ready` in the reverse direction.
+
+ - The data and user fields may be bundled in aggregate types named
+   `<stream-name>_data_type` and `<stream-name>_user_type` respectively. The
+   `data` signal becomes an array of `<stream-name>_data_type`s from 0 to
+   \\(N - 1\\) if \\(N > 1\\) or this is otherwise desirable.
```

Comment:
or this is otherwise desirable. -> or when otherwise desirable.

Reply:
Fixed in 129bbb9c